akka-projection-grpc/src/main/scala/akka/projection/grpc/replication/javadsl/Replication.scala

```diff
@@ -63,24 +64,58 @@ trait Replication[Command] {
 @ApiMayChange
 object Replication {
 
+  trait ReplicatedBehaviorFactory[Command, Event, State] {
```

Comment:
This kind of solution would mean that we need another one for timers, but maybe that is fine?

I think it should go in the replicated entity APIs in core Akka rather than here though

Reply:
Can be used with timers as well. Works with any behaviour composition. So can do this too:

```scala
Behaviors.setup[Command] { context =>
  Behaviors.withTimers { timers =>
    replicationSetup { replicationContext =>
      new ShoppingCart(context, replicationContext, timers).behavior()
    }
  }
}
```

This is what the replication setup variant takes:

```scala
((ReplicationContext => EventSourcedBehavior[Command, Event, State]) => Behavior[Command]) => Behavior[Command]
```

where you pass the replicated behaviour factory function to the given function (which creates the replication context and applies external replication) and returns a behaviour which can then be composed within another behaviour factory function.

Seems a little difficult to understand, although with some explicit types and better naming, could be ok. Maybe also calling it "setup" doesn't make it clear right now.

Reply:
Aha, I misread it, good if it works for arbitrary composition so we don't have to cover any possible use case.

I wonder if we could make the ReplicatedEntity.externalReplication less picky somehow, the reason why it requires `EventSourcedBehavior` to `externalReplication` is that it needs to set the flag `.withReplication(context)`, which is internal. Maybe we can pass open upp that and pass that responsibility to the user instead?

Reply:
Yes, wondered the same about just opening up `withReplication`. Although not sure if needing to apply that yourself is that much different from needing to use a provided factory.

I've tidied up the factory approach. Named the SAM class `ReplicatedBehaviors` with a `setup` method, so it looks similar to other behaviour factories, but provided to you in the superfactory instead of static methods. Also made this the only way, rather than two different factory approaches. See https://github.com/akka/akka-projection/pull/769/commits/343ae7178411e1efd9d6a187b1c23c8b24643b71.

Looks like this in the shopping cart now:

```scala
def apply(replicatedBehaviors: ReplicatedBehaviors[Command, Event, State]): Behavior[Command] = {
  Behaviors.setup[Command] { context =>
    replicatedBehaviors.setup { replicationContext =>
      new ShoppingCart(context, replicationContext).behavior()
    }
  }
}
```